File: 15threeSum/threeSum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        std::invoke_result_t<decltype(&Solution::threeSum), Solution, decltype(nums)> result;
        std::sort(std::begin(nums), std::end(nums));
        if (std::size(nums) < 3) return result;
        for (auto index{std::cbegin(nums)}; index != std::prev(std::cend(nums), 2);)
        {
            const auto num{*index};
            if (*index > 0) break;
            auto left{std::next(index)}, right{--std::cend(nums)};
            while (left != right)
            {
                const auto numsLeft{*left}, numsRight{*right};
                const auto total{num + numsLeft + numsRight};
                if (total == 0) result.emplace_back(std::vector{num, numsLeft, numsRight});
                if (total <= 0) while (left != right and numsLeft == *left) ++left;
                if (total >= 0) while (left != right and numsRight == *right) --right;
            }
            while (num == *index && index != std::prev(std::cend(nums), 2)) ++index;
        }
        return result;
    }
};

File: 15threeSum/threeSum.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function(nums) {
    nums.sort((a,b) => a - b)
    const result = []
    for (let index = 0; index < nums.length - 2;)
    {
        const num = nums[index]
        if (num > 0) break
        let left = index + 1, right = nums.length - 1
        while (!Object.is(left, right))
        {
            const numsLeft = nums[left], numsRight = nums[right]
            const total = num + numsLeft + numsRight
            if (Object.is(total, 0)) result.push([num, numsLeft, numsRight])
            if (total <= 0) while (!Object.is(left, right) && Object.is(numsLeft, nums[left])) ++left
            if (total >= 0) while (!Object.is(left, right) && Object.is(numsRight, nums[right])) --right
        }
        while (index < nums.length - 2 && Object.is(nums[index], num)) ++index
    }
    return result
};
